java/src/main/java/org/apache/arrow/gandiva/evaluator/NativeBuilder.java

```diff
@@ -16,26 +16,91 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+
 /**
  * This class is implemented in JNI. This provides the Java interface
  * to invoke functions in JNI
  */
 class NativeBuilder {
   private static final String LIBRARY_NAME = "gandiva_jni";
+  private static final String IRHELPERS_BC = "irhelpers.bc";
+
+  private static volatile NativeBuilder INSTANCE;
+
+  private final String byteCodeFilePath;
+
+  private NativeBuilder(String byteCodeFilePath) {
+    this.byteCodeFilePath = byteCodeFilePath;
+  }
+
+  static NativeBuilder getInstance() throws GandivaException {
+    if (INSTANCE == null) {
+      synchronized (NativeBuilder.class) {
+        if (INSTANCE == null) {
+          INSTANCE = setupInstance();
+        }
+      }
+    }
+    return INSTANCE;
+  }
+
+  String getByteCodeFilePath() {
+    return byteCodeFilePath;
+  }
 
-  static {
+  private static NativeBuilder setupInstance() throws GandivaException {
     try {
-      // TODO: Load the gandiva_jni lib dynamically from a jar
-      System.loadLibrary(LIBRARY_NAME);
-    } catch (UnsatisfiedLinkError unsatisfiedLinkError) {
-      String errMsg = "Unable to load native library: "
-              + LIBRARY_NAME
-              + " due to "
-              + unsatisfiedLinkError.toString();
-      System.err.println(errMsg);
-      // TODO: Need to handle this better. Gandiva cannot exit in case of a failure
-      System.exit(1);
+      String tempDir = System.getProperty("java.io.tmpdir");
+      loadGandivaLibraryFromJar(tempDir);
+      File byteCodeFile = moveFileFromJarToTemp(tempDir, IRHELPERS_BC);
+      return new NativeBuilder(byteCodeFile.getAbsolutePath());
+    } catch (IOException ioException) {
+      throw new GandivaException("unable to create native instance", ioException);
+    }
+  }
+
+  private static void loadGandivaLibraryFromJar(final String tmpDir)
+          throws IOException, GandivaException {
+    final String libraryToLoad = System.mapLibraryName(LIBRARY_NAME);
+    final File libraryFile = moveFileFromJarToTemp(tmpDir, libraryToLoad);
+    System.load(libraryFile.getAbsolutePath());
+  }
+
+
+  private static File moveFileFromJarToTemp(final String tmpDir, String libraryToLoad)
+          throws IOException, GandivaException {
+    final File temp = setupFile(tmpDir, libraryToLoad);
+    try (final InputStream is = NativeBuilder.class.getClassLoader()
+            .getResourceAsStream(libraryToLoad)) {
+      if (is == null) {
+        throw new GandivaException(libraryToLoad + " was not found inside JAR.");
+      } else {
+        Files.copy(is, temp.toPath(), StandardCopyOption.REPLACE_EXISTING);
+      }
+    }
+    return temp;
+  }
+
+  private static File setupFile(String tmpDir, String libraryToLoad)
+          throws IOException, GandivaException {
+    final File temp = new File(tmpDir, libraryToLoad);
+    if (temp.exists() && !temp.delete()) {
```

Comment:
Does this mean that only program can use Gandiva Java on a machine?

Reply:
good catch Vivek! fixed it.